### PR TITLE
Issue/7907 -> master

### PR DESCRIPTION
**Pull request by BeneBalazs**

### CODE REVIEW

File: web/modules/custom/ingredient_validation/src/Plugin/Validation/Constraint/IngredientConstraint.php [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+class IngredientConstraint extends Constraint {
+  /**
+   * Message shown when validation fails.
+   *
+   * @var string
+   */
+  public $message = 'Product creation failed: Ingredients must be in the correct form (Amount Unit Ingredient) and one in a line.';
+
+  /*
+   * {@inheritdoc}
```
Comment (gkurucz) [inline web/modules/custom/ingredient_validation/src/Plugin/Validation/Constraint/IngredientConstraint.php]: Some leftover doc comment? :slightly_smiling_face:

File: web/modules/custom/ingredient_validation/src/Plugin/Validation/Constraint/IngredientConstraintValidator.php [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+
+namespace Drupal\ingredient_validation\Plugin\Validation\Constraint;
+
+use Symfony\Component\Validator\Constraint;
+use Symfony\Component\Validator\ConstraintValidator;
+
+/**
+ * Validates the Ingredients field on the Product CT.
+ */
+class IngredientConstraintValidator extends ConstraintValidator {
```
Comment (gkurucz) [inline web/modules/custom/ingredient_validation/src/Plugin/Validation/Constraint/IngredientConstraintValidator.php]: In general it's a good practice to make a class final if you don't want other classes to extend this. Can you think of a case when this class would be inherited? It's always easier to 'un-final' a class than have unwanted inheritances.

https://www.php.net/manual/en/language.oop5.final.php

File: web/modules/custom/ingredient_validation/src/Plugin/Validation/Constraint/IngredientConstraintValidator.php [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+    $ingredients = $value->value;
+    $lines = explode('\n', strval($ingredients));
+    $pattern = "/^\d\s\w+\s\w+\s\w+$/";
+    foreach ($lines as $line) {
+      if (!preg_match($pattern, $line)) {
+        $this->context->addViolation($constraint->message);
+        return;
+      }
+    }
+
```
Comment (gkurucz) [inline web/modules/custom/ingredient_validation/src/Plugin/Validation/Constraint/IngredientConstraintValidator.php]: I think this line is unnecessary.

File: web/modules/custom/ingredient_validation/src/Plugin/Validation/Constraint/IngredientConstraintValidator.php [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+ * Validates the Ingredients field on the Product CT.
+ */
+class IngredientConstraintValidator extends ConstraintValidator {
+
+  /**
+   * {@inheritdoc}
+   */
+  public function validate($value, Constraint $constraint) {
+    $ingredients = $value->value;
+    $lines = explode('\n', strval($ingredients));
```
Comment (gkurucz) [inline web/modules/custom/ingredient_validation/src/Plugin/Validation/Constraint/IngredientConstraintValidator.php]: No need to store `$ingredients` in a variable. You can call `$value->value` in `strval()`.

**Review by gkurucz — CHANGES_REQUESTED** [review]

File: web/modules/custom/ingredient_validation/ingredient_validation.info.yml [hunk 5]
```diff
@@ -0,0 +1,4 @@
+name: ingredient_validation
```
Comment (mxr576) [inline web/modules/custom/ingredient_validation/ingredient_validation.info.yml]: This should be a human readable name

File: web/modules/custom/ingredient_validation/ingredient_validation.info.yml [hunk 6]
```diff
@@ -0,0 +1,4 @@
...
+type: module
+description: Validate the Ingredient field on the Product content type
+core: 8.x
```
Comment (mxr576) [inline web/modules/custom/ingredient_validation/ingredient_validation.info.yml]: Missing dependency to the node module.

File: web/modules/custom/ingredient_validation/src/Plugin/Validation/Constraint/IngredientConstraint.php [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+
+use Symfony\Component\Validator\Constraint;
+
+/**
+ * Make sure that the ingredients are written correctly.
+ *
+ * @Constraint(
+ * id = "ValidIngredient",
+ * label = @Translation("Prevent Product CT creation
+ * if it isn't matched with the required formula", context = "Validation"),
```
Comment (mxr576) [inline web/modules/custom/ingredient_validation/src/Plugin/Validation/Constraint/IngredientConstraint.php]: This should be a single line commit. a short human readable label for the validator.

File: web/modules/custom/ingredient_validation/src/Plugin/Validation/Constraint/IngredientConstraint.php [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+ * type = "string"
+ * )
+ */
+class IngredientConstraint extends Constraint {
```
Comment (mxr576) [inline web/modules/custom/ingredient_validation/src/Plugin/Validation/Constraint/IngredientConstraint.php]: Make the class final.

File: web/modules/custom/ingredient_validation/src/Plugin/Validation/Constraint/IngredientConstraintValidator.php [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+
+/**
+ * Validates the Ingredients field on the Product CT.
+ */
+class IngredientConstraintValidator extends ConstraintValidator {
+
+  /**
+   * {@inheritdoc}
+   */
+  public function validate($value, Constraint $constraint) {
```
Comment (mxr576) [inline web/modules/custom/ingredient_validation/src/Plugin/Validation/Constraint/IngredientConstraintValidator.php]: Missing return type hint.

File: web/modules/custom/ingredient_validation/src/Plugin/Validation/Constraint/IngredientConstraintValidator.php (hunk 4 above)
Comment (mxr576) [inline web/modules/custom/ingredient_validation/src/Plugin/Validation/Constraint/IngredientConstraintValidator.php]: Also use PHP_EOL instead of `\n`

**Review by mxr576 — CHANGES_REQUESTED** [review]

**Review by gkurucz — DISMISSED** [review]

File: web/modules/custom/ingredient_validation/src/Plugin/Validation/Constraint/IngredientConstraintValidator.php [hunk 10]
```diff
@@ -0,0 +1,29 @@
...
+/**
+ * Validates the Ingredients field on the Product CT.
+ */
+final class IngredientConstraintValidator extends ConstraintValidator {
+
+  /**
+   * {@inheritdoc}
+   */
+  public function validate($value, Constraint $constraint): void {
+    $lines = explode(PHP_EOL, strval($value->value));
```
Comment (lekob) [inline web/modules/custom/ingredient_validation/src/Plugin/Validation/Constraint/IngredientConstraintValidator.php]:
```suggestion
    $lines = explode(PHP_EOL, (string) $value->value);
```
Maybe casting to string would be better.

The field this contraint and validator are attached to is a multi-value field. The ->value property accessor will only return the first value of the multi-value field. You have two choices here:

1. Take into account that you are validating a multi-value field.
2. Change the field to be a single-value field (discuss this option with your team, there might be a valid reason for using a multi-value field here). If the field already has multiple items in it on the production server... well... then option 1 is the less painful (and only practical) path.

Comment (BeneBalazs) [reply web/modules/custom/ingredient_validation/src/Plugin/Validation/Constraint/IngredientConstraintValidator.php]: We've already agreed on that we'll have one field_ingredient and Gábor Törtei will change this in a PR later.

File: web/modules/custom/ingredient_validation/ingredient_validation.info.yml [hunk 11]
```diff
@@ -0,0 +1,7 @@
+name: Ingredient Validation
+type: module
+description: Validate the Ingredient field on the Product content type
+core: 8.x
```
Comment (mxr576) [inline web/modules/custom/ingredient_validation/ingredient_validation.info.yml]: Use `core_version_requirement` key instead.

https://www.drupal.org/node/3070687

**Review by mxr576 — CHANGES_REQUESTED** [review]

File: web/modules/custom/ingredient_validation/src/Plugin/Validation/Constraint/IngredientConstraint.php [hunk 12]
```diff
@@ -0,0 +1,26 @@
...
+namespace Drupal\ingredient_validation\Plugin\Validation\Constraint;
+
+use Symfony\Component\Validator\Constraint;
+
+/**
+ * Make sure that the ingredients are written correctly.
+ *
+ * @Constraint(
+ * id = "ValidIngredient",
+ * label = @Translation(" Not valid form of ingredients."),
```
Comment (mxr576) [inline web/modules/custom/ingredient_validation/src/Plugin/Validation/Constraint/IngredientConstraint.php]: s/ Not/Not/g

Comment (mxr576) [reply web/modules/custom/ingredient_validation/src/Plugin/Validation/Constraint/IngredientConstraint.php]: Plus unnecessary full-stop.

**Review by mxr576 — DISMISSED** [review]

File: web/modules/custom/ingredient_validation/ingredient_validation.info.yml [hunk 13]
```diff
@@ -0,0 +1,7 @@
+name: Ingredient Validation
+type: module
+description: Validate the Ingredient field on the Product content type
+core_version_requirement: ^8
```
Comment (lekob) [inline web/modules/custom/ingredient_validation/ingredient_validation.info.yml]:
```suggestion
core_version_requirement: ^8.8
```

**Review by lekob — APPROVED** [review]

**Review by gkurucz — APPROVED** [review]

**Pull request merged**